### PR TITLE
add an empty tab group for app tools launch configuration

### LINKED ISSUE #435: Define tab group for App Engine Standard Launch configuration type
image

discovered as part of #429

### PR DESCRIPTION
**Pull request by elharo**
fix #435 @akerekes

### CODE REVIEW

**Comment by elharo** [comment]
If you're happy with this, and I'm OOO you can go ahead and merge it. If not, you can pull down the branch and fix it up.

**Comment by elharo** [comment]
!ENTRY org.eclipse.equinox.preferences 4 2 2016-07-22 22:02:05.217
!MESSAGE Problems occurred when invoking code from plug-in: "org.eclipse.equinox.preferences".
!STACK 0
java.lang.IllegalStateException: Workbench has not been created yet.
	at org.eclipse.ui.PlatformUI.getWorkbench(PlatformUI.java:93)
	at org.eclipse.debug.internal.ui.DebugUIPreferenceInitializer.setThemeBasedPreferences(DebugUIPreferenceInitializer.java:208)

**Comment by elharo** [comment]
Passes locally, investigating

**Comment by elharo** [comment]
OK, now fails in a new way; still only on Travis though:

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.474 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.appengine.localserver.ui.AppToolsTabGroupTest
testCreateTabs(com.google.cloud.tools.eclipse.appengine.localserver.ui.AppToolsTabGroupTest)  Time elapsed: 0.466 sec  <<< ERROR!
org.eclipse.e4.core.di.InjectionException: java.lang.NoClassDefFoundError: javax/annotation/PostConstruct
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:389)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:352)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:344)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:160)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at org.eclipse.e4.core.internal.di.InjectorImpl.inject(InjectorImpl.java:121)
	at org.eclipse.e4.core.internal.di.InjectorImpl.internalMake(InjectorImpl.java:345)
	at org.eclipse.e4.core.internal.di.InjectorImpl.make(InjectorImpl.java:264)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.make(ContextInjectionFactory.java:162)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.createDefaultHeadlessContext(E4Application.java:508)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.createDefaultContext(E4Application.java:522)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.createE4Workbench(E4Application.java:203)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:628)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:606)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at com.google.cloud.tools.eclipse.appengine.localserver.ui.AppToolsTabGroupTest.createWorkbench(AppToolsTabGroupTest.java:20)

**Comment by chanseokoh** [comment]
`org.eclipse.e4.core.di.InjectionException: java.lang.NoClassDefFoundError: javax/annotation/PostConstruct`

This looks familiar. May not be the same issue, but the reason for this in the past was that `javax.annotation.*` was split into two separate jars, one from FindBugs and the other from the usual javax.annotation jar.

**Comment by elharo** [comment]
Do you remember how we fixed that or in which PR that showed up?

**Comment by elharo** [comment]
Looks like it was
https://github.com/GoogleCloudPlatform/app-tools-lib-for-java/pull/126
in app-tools-lib-for-java
I wonder if a similar fix would work here?

**Comment by briandealwis** [comment]
-1 to starting the workbench.  Rather configure the `tycho-surefire-plugin` to `useUIHarness` and product to `org.eclipse.sdk.ide`.  See [this commit](https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/437/files#diff-15415183ecf7a3a3cf9743279b916073R22) on the swtbot tests for an example.

**Comment by elharo** [comment]
Honestly, looking at the code I can't see why any of this is happening. It should be a straight-forward unit test without any Eclipse framework.

**Comment by briandealwis** [comment]
> @elharo: Passes locally, investigating
> @elharo: Honestly, looking at the code I can't see why any of this is happening. It should be a straight-forward unit test without any Eclipse framework.

Tycho only runs with the minimum bundles required to properly resolve dependencies.  This is unlike PDE's launchers, like the JUnit Plug-in Test, which normally runs with all plugins.

I guess the `org.eclipse.debug.*` bundles assume there's a minimal workbench.

I'm hoping the `useUIHarness` and specifying the `product` will be sufficient.

The [PostConstruct problem](https://wiki.eclipse.org/Eclipse4/RCP/FAQ#Why_aren.27t_my_.40Inject-able.2F.40PostConstruct_methods_being_injected.3F) is usually due to the `javax.annotation` bundle not being found.  Eclipse provides a bundle containing the JSR330 `javax.annotation` annotations for Java5 that exports package with a version of 1.0 and 1.1.  It's  possible there's some no interference with the JSR305 @Nullable/@Nonnull annotations as the JSR305 jar is a bundle (`org.jsr305`) that exports the package `javax.annotation` at version 3.0.  That would satisfy E4's `javax.annotation` imports for version 1.0, and could cause problems.  But I don't think we have anything depending on `org.jsr305` (app-tools-lib no longer has a runtime dependency).

**Comment by briandealwis** [comment]
LGTM

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/pom.xml [hunk 1]
```diff
@@ -17,6 +17,18 @@
       <plugin>
         <groupId>org.eclipse.tycho</groupId>
         <artifactId>tycho-surefire-plugin</artifactId>
+         <configuration>
+           <useUIHarness>true</useUIHarness>
+           <useUIThread>false</useUIThread>
+           <product>org.eclipse.sdk.ide</product>
+           <!-- <application>org.eclipse.ui.ide.workbench</application> -->
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/pom.xml]: Do we need this line?

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/AppToolsTabGroup.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.google.cloud.tools.eclipse.appengine.localserver.ui;
+
+import org.eclipse.debug.ui.AbstractLaunchConfigurationTabGroup;
+import org.eclipse.debug.ui.ILaunchConfigurationDialog;
+import org.eclipse.debug.ui.ILaunchConfigurationTab;
+
+public class AppToolsTabGroup extends AbstractLaunchConfigurationTabGroup {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/AppToolsTabGroup.java]: I'm not sure using AppTools in the name is good. This is for local App Engine devserver launch config. It shouldn't refer to implementation details in my opinion (now that we got rid of the other runtimes).

**Pull request merged**